### PR TITLE
fix(calling): detectNetwork change

### PR DESCRIPTION
**Pull request by karthikraji2020**
# COMPLETES #[< SPARK-458913 >](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-458913)

## This pull request addresses

We’re calling detectNetworkChange method from calling the client’s [init function](https://github.com/webex/webex-js-sdk/blob/next/packages/calling/src/CallingClient/CallingClient.ts#L123) after we create line dictionary by initialising line instance. [detectNetworkChange()](https://github.com/webex/webex-js-sdk/blob/next/packages/calling/src/CallingClient/CallingClient.ts#L159) checks if we have any connection or mercury flap and then calls handleConnectionRestoration() once mercury is up again which internally tries to restore registration or trigger a new registration if not done already.

 

Now, What if the external user of the line hasn’t triggered the line.register() already and mercury connection flaps before it. detectNetworkChange will eventually try to [create the registration](https://github.com/webex/webex-js-sdk/blob/next/packages/calling/src/CallingClient/CallingClient.ts#L184) even when registration didn’t happen for the first time.
 
Acceptance criteria:

Registration should not happen after the mercury connection is up again when it was not triggered even once.

## by making the following changes

We have introduced a new state: IDLE. Lines will be in IDLE state initially and will change to Active once registered. It will only be set back to IDLE if deregistration is done on purpose. In the detetctNetworkChange function, we have added a condition to check for line status. For cases where the line is in an IDLE state, handleConnectionRestorion will not be triggered and we also set the retry flag back to false.

### Change Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

Testcase 1: Triggered mercury flap before initial registration. registered the device after the connection was back up.
Testcase 2: Triggered mercury flap after the device is in registered state and restoration takes place.
Testcase 3: Triggered deregistration manually, mercury flap occurs and restoration flow isn't triggered.

Logs: [Mercury_Flap_After_Registration.log](https://github.com/webex/webex-js-sdk/files/13236960/Mercury_Flap_After_Registration.log)
[Mercury_Flap_Before_Initial_Registration.log](https://github.com/webex/webex-js-sdk/files/13236962/Mercury_Flap_Before_Initial_Registration.log)
[Uploading Mercury_Flap_After_Deregistration.log…]()

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/calling/src/CallingClient/CallingClient.ts [hunk 1]
```diff
@@ -142,7 +142,7 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
         retry = true;
       }
 
-      if (retry && this.webex.internal.mercury.connected) {
+      if (retry && this.webex.internal.mercury.connected && line.mobiusDeviceId) {
```
Comment (Kesari3008) [inline packages/calling/src/CallingClient/CallingClient.ts:145]: Is there a possibility to add a UT for this by simulating this scenario? Also there maybe a possibility to test this manually also, have you tried to test it?

**Review by Kesari3008 — CHANGES_REQUESTED** [review]

File: packages/calling/src/CallingClient/CallingClient.ts (hunk 1 above)
Comment (karthikraji2020) [reply packages/calling/src/CallingClient/CallingClient.ts:145]: ack: updated

File: packages/calling/src/CallingClient/CallingClient.test.ts [hunk 2]
```diff
@@ -509,6 +509,48 @@ describe('CallingClient Tests', () => {
       expect(registerSpy).not.toBeCalled();
     });
 
+    it('Simulate a network flap before initial registration is done should not trigger re-verify registration: Restore Failure', async () => {
+      expect(line.getRegistrationStatus()).toEqual(MobiusStatus.ACTIVE);
+
+      reg.deregister();
+      line.registration.deregister();
```
Comment (sreenara) [inline packages/calling/src/CallingClient/CallingClient.test.ts:516]: Isn't reg = line.registration ? Why are we called line.registration.deregister() after this?
Also, why are we not doing line.deregister() directly?

**Review by sreenara — CHANGES_REQUESTED** [review]
Have a comment in the UT

File: packages/calling/src/CallingClient/CallingClient.test.ts (hunk 2 above)
Comment (karthikraji2020) [reply packages/calling/src/CallingClient/CallingClient.test.ts:516]: Ref:[/CallingClient/line/index.ts#L148]  (https://github.com/webex/webex-js-sdk/blob/a57a4decebcf721da6495c398b566e9b803fb6c3/packages/calling/src/CallingClient/line/index.ts#L148)
 within the line Object we have `registration` module..
 
 [the way we are accessing the registration modules method] (https://github.com/webex/webex-js-sdk/blob/a57a4decebcf721da6495c398b566e9b803fb6c3/packages/calling/src/CallingClient/CallingClient.test.ts#L377)

Comment (karthikraji2020) [reply packages/calling/src/CallingClient/CallingClient.test.ts:516]: > Isn't reg = line.registration ? Why are we called line.registration.deregister() after this? Also, why are we not doing line.deregister() directly?

Ref:[/CallingClient/line/index.ts#L148] (https://github.com/webex/webex-js-sdk/blob/a57a4decebcf721da6495c398b566e9b803fb6c3/packages/calling/src/CallingClient/line/index.ts#L148C1-L150C4)
 within the line Object we have `registration` module..
 
 [the way we are accessing the registration modules method](https://github.com/webex/webex-js-sdk/blob/a57a4decebcf721da6495c398b566e9b803fb6c3/packages/calling/src/CallingClient/CallingClient.test.ts#L377)

Comment (sreenara) [reply packages/calling/src/CallingClient/CallingClient.test.ts:516]: I still don't understand why we are doing it this way.
Why can't we just have 1 line saying `line.deregister()`?

Comment (meta-dipanshu-sharma) [reply packages/calling/src/CallingClient/CallingClient.test.ts:516]: @karthikraji2020

> so to access the deregister() we are using line.registration.deregister()

here, `reg == line.registration` as Sreekanth mentioned above so no need to separately call `line.registration.deregister()` to invoke deregister(). `reg.deregister()` will invoke the same function.

Use either of below not both,
1. `reg.deregister()`
2. `line.registration.deregister()`

Comment (karthikraji2020) [reply packages/calling/src/CallingClient/CallingClient.test.ts]: ack: updated

**Review by sreenara — APPROVED** [review]

File: packages/calling/src/CallingClient/CallingClient.test.ts [hunk 3]
```diff
@@ -509,6 +509,47 @@ describe('CallingClient Tests', () => {
...
+      webex.internal.mercury.connected = true;
+      jest.advanceTimersByTime(NETWORK_FLAP_TIMEOUT + 500);
+
+      await flushPromises();
+
+      /* We should be detecting the network recovery */
+
+      expect(logSpy).not.toBeCalledWith(
+        'Mercury connection is up again, Re-registering with Mobius',
+        {file: REGISTRATION_FILE, method: 'handleConnectionRestoration'}
```
Comment (Kesari3008) [inline packages/calling/src/CallingClient/CallingClient.test.ts:541]: Is it possible to spy on the handleConnectionRestoration method and check that it not getting invoked?

File: packages/calling/src/CallingClient/CallingClient.test.ts [hunk 4]
```diff
@@ -509,6 +509,47 @@ describe('CallingClient Tests', () => {
       expect(registerSpy).not.toBeCalled();
     });
 
+    it('Simulate a network flap before initial registration is done should not trigger re-verify registration: Restore Failure', async () => {
+      expect(line.getRegistrationStatus()).toEqual(MobiusStatus.ACTIVE);
+
+      reg.deregister();
+      line.mobiusDeviceId = undefined;
```
Comment (Kesari3008) [inline packages/calling/src/CallingClient/CallingClient.test.ts:516]: We are specifically setting mobiusDeviceId as undefined and deregistering the line. This scenario should be where there were no registrations at all

**Review by Kesari3008 — APPROVED** [review]
Approving but UT apporach needs discussing

File: packages/calling/src/CallingClient/registration/register.ts [hunk 5]
```diff
@@ -151,6 +150,7 @@ export class Registration implements IRegistration {
       },
     });
 
+    this.setStatus(RegistrationStatus.INACTIVE);
```
Comment (sreenara) [inline packages/calling/src/CallingClient/registration/register.ts:153]: Wouldn't it be better to move this to IDLE? We are deleting the registration on purpose here, so ideally we (developer) already know that the registration is inactive.
I feel INACTIVE is generally a state to be used when the registration has been deleted from the mobius side or we receive an error in registration.

File: packages/calling/src/CallingClient/registration/register.ts [hunk 6]
```diff
@@ -592,7 +592,7 @@ export class Registration implements IRegistration {
     for (const url of servers) {
       try {
         abort = false;
-        this.registrationStatus = MobiusStatus.DEFAULT;
+        this.registrationStatus = RegistrationStatus.INACTIVE;
```
Comment (sreenara) [inline packages/calling/src/CallingClient/registration/register.ts:595]: Should this be INACTIVE or IDLE?

File: packages/calling/src/CallingClient/registration/register.ts [hunk 7]
```diff
@@ -760,7 +760,7 @@ export class Registration implements IRegistration {
     }
 
     this.clearKeepaliveTimer();
-    this.setStatus(MobiusStatus.DEFAULT);
+    this.setStatus(RegistrationStatus.INACTIVE);
```
Comment (sreenara) [inline packages/calling/src/CallingClient/registration/register.ts:763]: Same question here as in deleteRegistration

File: packages/calling/src/CallingClient/CallingClient.ts [hunk 8]
```diff
@@ -162,7 +163,11 @@ export class CallingClient extends Eventing<CallingClientEventTypes> implements
         retry = true;
       }
 
-      if (retry && this.webex.internal.mercury.connected) {
+      if (
+        retry &&
+        this.webex.internal.mercury.connected &&
+        line.status !== RegistrationStatus.IDLE
+      ) {
         retry = await line.registration.handleConnectionRestoration(retry);
```
Comment (sreenara) [inline packages/calling/src/CallingClient/CallingClient.ts:171]: One comment for the log inside handleConnectionRestoration. Can we change it to "Mercury connection is up again, re-registering with Webex Calling if needed."

File: packages/calling/src/CallingClient/line/index.ts [hunk 9]
```diff
@@ -118,7 +118,7 @@ export default class Line extends Eventing<LineEventTypes> implements ILine {
       logLevel
     );
 
-    this.registration.setStatus(MobiusStatus.DEFAULT);
+    this.registration.setStatus(RegistrationStatus.INACTIVE);
```
Comment (sreenara) [inline packages/calling/src/CallingClient/line/index.ts:121]: Shouldn't this be IDLE?

File: packages/calling/src/CallingClient/line/index.ts [hunk 10]
```diff
@@ -189,6 +190,7 @@ export default class Line extends Eventing<LineEventTypes> implements ILine {
       case LINE_EVENTS.UNREGISTERED:
       case LINE_EVENTS.RECONNECTED:
       case LINE_EVENTS.RECONNECTING:
+        this.status = this.registration.getStatus();
```
Comment (sreenara) [inline packages/calling/src/CallingClient/line/index.ts:193]: Why did we need this?

**Review by sreenara — CHANGES_REQUESTED** [review]
Please see the comments. Tests also will require changes.

File: packages/calling/src/CallingClient/registration/register.ts (hunk 5 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/registration/register.ts:153]: As per our discussion, IDLE status is removed for now. We only have Active and Inactive.

File: packages/calling/src/CallingClient/registration/register.ts (hunk 6 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/registration/register.ts:595]: Same comment as above

File: packages/calling/src/CallingClient/registration/register.ts (hunk 7 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/registration/register.ts:763]: Here it needed to be INACTIVE only

File: packages/calling/src/CallingClient/CallingClient.ts (hunk 8 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.ts:171]: Updated

File: packages/calling/src/CallingClient/line/index.ts (hunk 10 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/line/index.ts:193]: Removed the status variable itself from the line object as discussed

**Comment by meta-dipanshu-sharma** [comment]
@Kesari3008 I'm reviewing the PR but just have one question that what was the issue with the earlier fix when we added check for mobiusDeviceId on line when connection is up again?

File: packages/calling/src/CallingClient/line/index.ts [hunk 11]
```diff
@@ -131,7 +127,9 @@ export default class Line extends Eventing<LineEventTypes> implements ILine {
    */
   public async register() {
     await this.#mutex.runExclusive(async () => {
-      this.registration.setStatus(MobiusStatus.DEFAULT);
+      if (this.registration.getStatus() !== RegistrationStatus.IDLE) {
```
Comment (sreenara) [inline packages/calling/src/CallingClient/line/index.ts:130]: let's remove this code and check what happens

File: packages/calling/src/CallingClient/line/index.ts [hunk 12]
```diff
@@ -118,7 +114,7 @@ export default class Line extends Eventing<LineEventTypes> implements ILine {
       logLevel
     );
 
-    this.registration.setStatus(MobiusStatus.DEFAULT);
+    this.registration.setStatus(RegistrationStatus.IDLE);
```
Comment (sreenara) [inline packages/calling/src/CallingClient/line/index.ts:117]: this shouldn't be needed

**Review by sreenara — CHANGES_REQUESTED** [review]
Please check the comments

File: packages/calling/src/CallingClient/registration/register.ts (hunk 5 above)
Comment (sreenara) [reply packages/calling/src/CallingClient/registration/register.ts:158]: Clarified now.

File: packages/calling/src/CallingClient/registration/register.ts (hunk 6 above)
Comment (sreenara) [reply packages/calling/src/CallingClient/registration/register.ts:600]: We can leave this as is

File: packages/calling/src/CallingClient/registration/register.ts (hunk 7 above)
Comment (sreenara) [reply packages/calling/src/CallingClient/registration/register.ts:767]: As discussed, it needs to be IDLE here for a purposeful deregistration by the developer

**Review by sreenara — CHANGES_REQUESTED** [review]
Resolved all comments that we discussed where no changes are needed.
I've only left the comments where the changes are required.

Also, for the failure registration cases, I think we should mark the status as INACTIVE in [this condition](https://github.com/webex/webex-js-sdk/blob/d0744a57a8d61eb2795097474aa2d528414665cb/packages/calling/src/CallingClient/registration/register.ts#L655). What do you think?

**Comment by Kesari3008** [comment]
> @Kesari3008 I'm reviewing the PR but just have one question that what was the issue with the earlier fix when we added check for mobiusDeviceId on line when connection is up again? PR description still explains the old fix so asking here.

@meta-dipanshu-sharma I have updated the description please check. The reason why we cannot use deviceID in this condition because once we trigger a manual registration after the mercury flap, we could see restoration flow also kicking in and also device is never deleted from SDK. So deviceID will also be present in scenarios where we do not want to trigger restoration.

File: packages/calling/src/CallingClient/registration/register.ts (hunk 7 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/registration/register.ts:767]: As we discussed yesterday, we cannot change the state to IDLE in deleteRegistration as this gets invoked in restoration and failback cases also

File: packages/calling/src/CallingClient/line/index.ts (hunk 11 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/line/index.ts]: this is removed

File: packages/calling/src/CallingClient/line/index.ts (hunk 12 above)
Comment (Kesari3008) [reply packages/calling/src/CallingClient/line/index.ts]: removed..sorry missed this comment

**Review by sreenara — APPROVED** [review]

**Pull request merged**